File: go/plugins.go
// plugins.go
//
// Ports the portable core of CircleAI.Plugins (IPlugin.cs + PluginContext.cs):
//
//	IPlugin / IPluginContext / IPluginEvents  -> Plugin / PluginContext / PluginEvents (ifaces)
//	PluginEvents (class)                       -> DefaultPluginEvents (thread-safe bus)
//	PluginEventNames (static)                  -> package constants
//	PluginContext (class)                      -> DefaultPluginContext
//	PermissionedPluginContext                  -> PermissionedPluginContext (+ Permissions)
//
// Out of the portable surface (host-side .NET plumbing, deliberately not ported —
// same call the Go tree makes for ServiceCollection / IHostedService / telemetry):
//   - PluginLoader (System.Runtime.Loader.AssemblyLoadContext reflection),
//   - PluginLifecycleService (IHostedService + IConfiguration + DI),
//   - PluginRegistry / PluginMarketplace (JSON-file persistence + File I/O).
// The stable surface plugins actually consume (event bus, logger, workspace
// path, permission gating) is ported in full and is pure in-memory.
//
// The C# IPluginContext exposes an ILogger; here it is the minimal PluginLogger
// seam (nil = no-op), matching the RuntimeLogger convention in companion_runtime.go.

package circleai

import (
	"strings"
	"sync"
)

// PluginLogger is the minimal logging seam an IPluginContext exposes. Ports the
// ILogger dependency (narrowed to what a plugin needs). A nil PluginLogger is a
// no-op — mirroring NullLogger.Instance.
type PluginLogger interface {
	// Log records a message at an informational level with optional args.
	Log(message string, args ...any)
}

// Plugin is the contract every CircleAI plugin implements. Ports IPlugin.
type Plugin interface {
	ID() string
	DisplayName() string
	Version() string
	// Initialize is called once at host startup with the plugin's context.
	Initialize(ctx PluginContext) error
	// Shutdown is called when the host is stopping or the plugin is unloaded.
	Shutdown() error
}

// PluginContext is the stable surface plugins are allowed to use. Ports
// IPluginContext. WorkspacePath returns "" when not set (C# nullable string).
type PluginContext interface {
	WorkspacePath() string
	Events() PluginEvents
	Logger() PluginLogger
}

// PluginEvents is the string-keyed event bus. Ports IPluginEvents. Subscribe
// returns an unsubscribe func in place of the C# IDisposable handle.
type PluginEvents interface {
	// Subscribe registers handler for eventName and returns an unsubscribe func.
	Subscribe(eventName string, handler func(payload any)) (unsubscribe func())
	// Raise fires an event to all subscribers (host-only API).
	Raise(eventName string, payload any)
}

// Well-known event names. Ports PluginEventNames.
const (
	// PluginEventWorkspaceLoaded — "workspace.loaded".
	PluginEventWorkspaceLoaded = "workspace.loaded"
	// PluginEventChatMessage — "chat.message".
	PluginEventChatMessage = "chat.message"
	// PluginEventModelLoaded — "model.loaded".
	PluginEventModelLoaded = "model.loaded"
	// PluginEventModelUnloaded — "model.unloaded".
	PluginEventModelUnloaded = "model.unloaded"
)

// DefaultPluginEvents is the thread-safe default event bus. Ports the
// PluginEvents class. The zero value is not usable — construct with
// NewDefaultPluginEvents.
//
// CONCURRENCY: Raise snapshots the handler slice for the event UNDER the lock and
// invokes callbacks OUTSIDE it, so a handler that (un)subscribes cannot deadlock
// the raiser. Handler panics are swallowed (an unhealthy plugin must not corrupt
// the host — matching the C# empty catch).
type DefaultPluginEvents struct {
	mu       sync.Mutex
	handlers map[string][]*pluginHandler
}

type pluginHandler struct {
	fn func(payload any)
}

// NewDefaultPluginEvents constructs an empty event bus.
func NewDefaultPluginEvents() *DefaultPluginEvents {
	return &DefaultPluginEvents{handlers: make(map[string][]*pluginHandler)}
}

// Subscribe registers handler for eventName and returns an idempotent
// unsubscribe func. Ports Subscribe. Panics if eventName is blank or handler is
// nil (mirrors the C# ArgumentException / ArgumentNullException).
func (e *DefaultPluginEvents) Subscribe(eventName string, handler func(payload any)) (unsubscribe func()) {
	if strings.TrimSpace(eventName) == "" {
		panic("eventName required")
	}
	if handler == nil {
		panic("handler must not be nil")
	}
	h := &pluginHandler{fn: handler}
	e.mu.Lock()
	e.handlers[eventName] = append(e.handlers[eventName], h)
	e.mu.Unlock()
	var once sync.Once
	return func() { once.Do(func() { e.unsubscribe(eventName, h) }) }
}

func (e *DefaultPluginEvents) unsubscribe(eventName string, h *pluginHandler) {
	e.mu.Lock()
	defer e.mu.Unlock()
	list := e.handlers[eventName]
	for i, x := range list {
		if x == h {
			e.handlers[eventName] = append(list[:i], list[i+1:]...)
			return
		}
	}
}

// Raise fires payload to all subscribers of eventName. Ports Raise.
func (e *DefaultPluginEvents) Raise(eventName string, payload any) {
	e.mu.Lock()
	list := e.handlers[eventName]
	snap := make([]*pluginHandler, len(list))
	copy(snap, list)
	e.mu.Unlock()
	for _, h := range snap {
		func() {
			defer func() { _ = recover() }()
			h.fn(payload)
		}()
	}
}

// DefaultPluginContext is the default IPluginContext. Ports the PluginContext
// class. Construct with NewDefaultPluginContext.
type DefaultPluginContext struct {
	workspacePath func() string
	events        PluginEvents
	logger        PluginLogger
}

// NewDefaultPluginContext constructs a context. workspacePathAccessor may be nil
// (WorkspacePath then returns ""). events must not be nil.
func NewDefaultPluginContext(workspacePathAccessor func() string, events PluginEvents, logger PluginLogger) *DefaultPluginContext {
	if events == nil {
		panic("events must not be nil")
	}
	if workspacePathAccessor == nil {
		workspacePathAccessor = func() string { return "" }
	}
	return &DefaultPluginContext{workspacePath: workspacePathAccessor, events: events, logger: logger}
}

// WorkspacePath returns the host-configured workspace path. Ports the
// WorkspacePath property.
func (c *DefaultPluginContext) WorkspacePath() string { return c.workspacePath() }

// Events returns the event bus. Ports the Events property.
func (c *DefaultPluginContext) Events() PluginEvents { return c.events }

// Logger returns the plugin logger. Ports the Logger property.
func (c *DefaultPluginContext) Logger() PluginLogger { return c.logger }

// Plugin permission constants. Ports PermissionedPluginContext.Permissions.
const (
	// PluginPermissionWorkspaceRead — "workspace.read".
	PluginPermissionWorkspaceRead = "workspace.read"
	// PluginPermissionWorkspaceWrite — "workspace.write".
	PluginPermissionWorkspaceWrite = "workspace.write"
	// PluginPermissionEventsSubscribe — "events.subscribe".
	PluginPermissionEventsSubscribe = "events.subscribe"
)

// PermissionedPluginContext wraps an inner context and gates capabilities by a
// granted-permission set. Ports PermissionedPluginContext. Construct with
// NewPermissionedPluginContext.
type PermissionedPluginContext struct {
	inner   PluginContext
	granted map[string]bool
	events  PluginEvents
}

// NewPermissionedPluginContext wraps inner, gating by grantedPermissions
// (case-insensitive). Without events.subscribe, Events returns a drop-on-the-floor
// bus; without workspace.read/write, WorkspacePath returns "". Panics if inner is
// nil.
func NewPermissionedPluginContext(inner PluginContext, grantedPermissions []string) *PermissionedPluginContext {
	if inner == nil {
		panic("inner must not be nil")
	}
	granted := make(map[string]bool, len(grantedPermissions))
	for _, p := range grantedPermissions {
		granted[strings.ToLower(p)] = true
	}
	var events PluginEvents
	if granted[PluginPermissionEventsSubscribe] {
		events = inner.Events()
	} else {
		events = silentPluginEvents{}
	}
	return &PermissionedPluginContext{inner: inner, granted: granted, events: events}
}

// WorkspacePath returns the inner path only if a workspace permission is granted.
// Ports the WorkspacePath property.
func (c *PermissionedPluginContext) WorkspacePath() string {
	if c.granted[PluginPermissionWorkspaceRead] || c.granted[PluginPermissionWorkspaceWrite] {
		return c.inner.WorkspacePath()
	}
	return ""
}

// Events returns the (possibly silent) event bus. Ports the Events property.
func (c *PermissionedPluginContext) Events() PluginEvents { return c.events }

// Logger returns the inner logger. Ports the Logger property.
func (c *PermissionedPluginContext) Logger() PluginLogger { return c.inner.Logger() }

// silentPluginEvents drops every event (permission-denied plugins). Ports the
// SilentEvents nested class.
type silentPluginEvents struct{}

func (silentPluginEvents) Subscribe(eventName string, handler func(payload any)) (unsubscribe func()) {
	return func() {}
}
func (silentPluginEvents) Raise(eventName string, payload any) {}

// Interface guards.
var (
	_ PluginEvents  = (*DefaultPluginEvents)(nil)
	_ PluginEvents  = silentPluginEvents{}
	_ PluginContext = (*DefaultPluginContext)(nil)
	_ PluginContext = (*PermissionedPluginContext)(nil)
)

File: c/src/plugins.c
/*
 * plugins.c — CircleAI.Plugins (C11 port).
 *
 * PluginEvents: string-keyed handler lists; Raise snapshots the matching list
 * before invoking so a handler may unsubscribe safely. PluginContext carries a
 * workspace accessor + events + opaque logger. PermissionedPluginContext gates
 * WorkspacePath + Events by a granted-permission set (a silent bus stands in
 * when events are denied).
 *
 * Pure C11 + libc. No pthreads.
 */

#include "circle_ai/plugins.h"
#include "board_common.h"

/* ── PluginEvents ───────────────────────────────────────────────────────── */

struct ca_plugin_subscription {
    char                *event_name; /* owned */
    ca_plugin_handler_fn handler;
    void                *ctx;
    ca_plugin_events_t  *owner;      /* borrowed */
    bool                 live;
};

struct ca_plugin_events {
    ca_plugin_subscription_t **subs; /* owned tokens */
    size_t                     count, cap;
    bool                       silent; /* SilentEvents: drop raises, noop tokens */
};

static ca_plugin_events_t *events_create_impl(bool silent) {
    ca_plugin_events_t *e =
        (ca_plugin_events_t *)calloc(1, sizeof(ca_plugin_events_t));
    if (e) e->silent = silent;
    return e;
}
ca_plugin_events_t *ca_plugin_events_create(void) {
    return events_create_impl(false);
}
void ca_plugin_events_destroy(ca_plugin_events_t *e) {
    if (!e) return;
    for (size_t i = 0; i < e->count; ++i) {
        free(e->subs[i]->event_name);
        free(e->subs[i]);
    }
    free(e->subs);
    free(e);
}

ca_plugin_subscription_t *ca_plugin_events_subscribe(ca_plugin_events_t *e,
                                                     const char *event_name,
                                                     ca_plugin_handler_fn handler,
                                                     void *ctx) {
    if (!e || cab_is_ws(event_name) || !handler) return NULL;
    if (e->silent) {
        /* SilentEvents.Subscribe -> NoopDisposable. Return a live-but-detached
         * token that unsubscribe treats as a no-op. */
        ca_plugin_subscription_t *nop =
            (ca_plugin_subscription_t *)calloc(1, sizeof(*nop));
        if (!nop) return NULL;
        nop->owner = e;
        nop->live = false; /* not registered */
        nop->event_name = cab_strdup(event_name);
        if (!nop->event_name) { free(nop); return NULL; }
        return nop;
    }
    ca_plugin_subscription_t *sub =
        (ca_plugin_subscription_t *)calloc(1, sizeof(*sub));
    if (!sub) return NULL;
    sub->event_name = cab_strdup(event_name);
    if (!sub->event_name) { free(sub); return NULL; }
    sub->handler = handler;
    sub->ctx = ctx;
    sub->owner = e;
    sub->live = true;
    if (e->count == e->cap) {
        size_t nc = e->cap ? e->cap * 2 : 4;
        void *n = realloc(e->subs, nc * sizeof(*e->subs));
        if (!n) { free(sub->event_name); free(sub); return NULL; }
        e->subs = (ca_plugin_subscription_t **)n;
        e->cap = nc;
    }
    e->subs[e->count++] = sub;
    return sub;
}

int ca_plugin_events_raise(ca_plugin_events_t *e, const char *event_name,
                           void *payload) {
    if (!e || !event_name || e->silent) return 0;
    /* Snapshot matching handlers first. */
    size_t match = 0;
    for (size_t i = 0; i < e->count; ++i)
        if (e->subs[i]->live && cab_ord_eq(e->subs[i]->event_name, event_name))
            match++;
    if (match == 0) return 0;

    ca_plugin_subscription_t **snap =
        (ca_plugin_subscription_t **)malloc(match * sizeof(*snap));
    if (!snap) return 0;
    size_t k = 0;
    for (size_t i = 0; i < e->count; ++i)
        if (e->subs[i]->live && cab_ord_eq(e->subs[i]->event_name, event_name))
            snap[k++] = e->subs[i];
    for (size_t i = 0; i < match; ++i)
        snap[i]->handler(snap[i]->ctx, payload);
    free(snap);
    return (int)match;
}

void ca_plugin_events_unsubscribe(ca_plugin_events_t *e,
                                  ca_plugin_subscription_t *sub) {
    if (!sub) return;
    if (!sub->live) {
        /* detached silent-bus token: just free it */
        free(sub->event_name);
        free(sub);
        return;
    }
    if (!e) return;
    for (size_t i = 0; i < e->count; ++i) {
        if (e->subs[i] == sub) {
            free(sub->event_name);
            free(sub);
            e->subs[i] = e->subs[e->count - 1];
            e->count--;
            return;
        }
    }
}

/* ── IPluginContext ─────────────────────────────────────────────────────── */

struct ca_plugin_context {
    ca_plugin_workspace_fn accessor;
    void                  *accessor_ctx;
    ca_plugin_events_t    *events;      /* borrowed (or owned silent bus) */
    void                  *logger;      /* borrowed opaque */
    bool                   owns_events; /* true when we made a silent bus */
    /* Permissioned gating (for the wrapper). */
    bool                   permissioned;
    bool                   allow_workspace; /* WorkspacePath exposed? */
    const ca_plugin_context_t *inner;    /* borrowed, for the wrapper */
};

ca_plugin_context_t *ca_plugin_context_create(ca_plugin_workspace_fn accessor,
                                              void *accessor_ctx,
                                              ca_plugin_events_t *events,
                                              void *logger) {
    if (!events || !logger) return NULL;
    ca_plugin_context_t *c =
        (ca_plugin_context_t *)calloc(1, sizeof(ca_plugin_context_t));
    if (!c) return NULL;
    c->accessor = accessor;
    c->accessor_ctx = accessor_ctx;
    c->events = events;
    c->logger = logger;
    return c;
}
void ca_plugin_context_destroy(ca_plugin_context_t *c) {
    if (!c) return;
    if (c->owns_events) ca_plugin_events_destroy(c->events);
    free(c);
}

const char *ca_plugin_context_workspace_path(const ca_plugin_context_t *c) {
    if (!c) return NULL;
    if (c->permissioned) {
        if (!c->allow_workspace) return NULL;
        return ca_plugin_context_workspace_path(c->inner);
    }
    if (!c->accessor) return NULL;
    return c->accessor(c->accessor_ctx);
}
ca_plugin_events_t *ca_plugin_context_events(const ca_plugin_context_t *c) {
    return c ? c->events : NULL;
}
void *ca_plugin_context_logger(const ca_plugin_context_t *c) {
    if (!c) return NULL;
    if (c->permissioned) return ca_plugin_context_logger(c->inner);
    return c->logger;
}

/* Case-insensitive membership in the granted set. */
static bool granted_has(const char *const *granted, size_t n, const char *perm) {
    for (size_t i = 0; i < n; ++i)
        if (cab_ci_eq(granted[i], perm)) return true;
    return false;
}

ca_plugin_context_t *ca_plugin_context_permissioned(
    const ca_plugin_context_t *inner, const char *const *granted,
    size_t granted_count) {
    if (!inner) return NULL;
    ca_plugin_context_t *c =
        (ca_plugin_context_t *)calloc(1, sizeof(ca_plugin_context_t));
    if (!c) return NULL;
    c->permissioned = true;
    c->inner = inner;
    c->logger = inner->logger;

    bool events_ok = granted_has(granted, granted_count, CA_PLUGIN_PERM_EVENTS_SUBSCRIBE);
    c->allow_workspace =
        granted_has(granted, granted_count, CA_PLUGIN_PERM_WORKSPACE_READ) ||
        granted_has(granted, granted_count, CA_PLUGIN_PERM_WORKSPACE_WRITE);

    if (events_ok) {
        c->events = inner->events;
        c->owns_events = false;
    } else {
        /* SilentEvents */
        c->events = events_create_impl(true);
        if (!c->events) { free(c); return NULL; }
        c->owns_events = true;
    }
    return c;
}

/* ── IPlugin dispatchers ────────────────────────────────────────────────── */

int ca_plugin_initialize(const ca_plugin_t *p, ca_plugin_context_t *context) {
    if (!p || !p->initialize || !context) return -1;
    return p->initialize(p->ctx, context);
}
int ca_plugin_shutdown(const ca_plugin_t *p) {
    if (!p || !p->shutdown) return -1;
    return p->shutdown(p->ctx);
}
